--- a/src/tl_templates/cuda/instruction/mma_block_scale.h
+++ b/src/tl_templates/cuda/instruction/mma_block_scale.h
@@ -38,8 +38,8 @@
     uint32_t b1, const float *c, uint32_t scale_a, uint32_t scale_b,
     uint16_t scale_a_byte_id = 0, uint16_t scale_a_thread_id = 0,
     uint16_t scale_b_byte_id = 0, uint16_t scale_b_thread_id = 0) {
-#if defined(CUTE_ARCH_MXF4NVF4_4X_UE4M3_MMA_ENABLED) &&                        \
-    defined(CUTLASS_ARCH_MMA_SM120A_ENABLED)
+#if defined(CUTE_ARCH_MXF4NVF4_4X_UE4M3_MMA_ENABLED) &&                        \
+    (defined(CUTLASS_ARCH_MMA_SM120A_ENABLED) || defined(CUTLASS_ARCH_MMA_SM121A_ENABLED))
   asm volatile(
       "mma.sync.aligned.m16n8k64.row.col.kind::mxf4nvf4.block_scale.scale_vec::"
       "4X.f32.e2m1.e2m1.f32.ue4m3 "
